### PR TITLE
Speed up RPyC connection

### LINKED ISSUE #1821: Improve RPyC connection performance
Currently (as of #1818) RPyC is being used as the main communication channel between local Modin and scaled-out to the cluster.
It does a lot of excessive calls, and each call costs at least as much time as ping to spawned cluster (because most of the calls need to be synchronous).

We need to make less calls and cache whatever is safe.

### PR DESCRIPTION
**Pull request by vnlitvinov**
## What do these changes do?
* Cache remote class types of known classes (RPyC by default reconstructs them upon each call)
* Use more efficient "batching" `.deliver()` which transfers all arguments in a single RPyC call instead of one call per argument
* Override inspecting certain classes, cache some getting attributes, hashes and strings for known static-ish items in `numpy`
* Obtain as local tuple the results of calling `DF.dtypes.items()` and `DF.dtypes.iteritems()`, as iterating over a remote connection introduces a lot of latency.

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #1821
- [ ] tests added and passing

### CODE REVIEW

File: modin/experimental/cloud/rpyc_proxy.py [hunk 1]
```diff
@@ -12,34 +12,157 @@
...
+        self._remote_dumps = None
+        self._remote_tuplize = None
+
+    def __wrap(self, local_obj):
+        while True:
+            # unwrap magic wrappers first
+            try:
+                local_obj = object.__getattribute__(local_obj, "__remote_end__")
+            except AttributeError:
+                break
```
Comment (anmyachev) [inline modin/experimental/cloud/rpyc_proxy.py:51]: You using a loop here, why?

File: modin/experimental/cloud/rpyc_proxy.py [hunk 2]
```diff
@@ -12,34 +12,157 @@
...
+        # do not pickle netrefs
+        if isinstance(local_obj, netref.BaseNetref) and local_obj.____conn__ is self:
```
Comment (anmyachev) [inline modin/experimental/cloud/rpyc_proxy.py:53]: Why is `if isinstance(local_obj, netref.BaseNetref):` not enough?

File: modin/experimental/cloud/rpyc_proxy.py [hunk 3]
```diff
@@ -353,9 +489,13 @@ class DeliveringMixin:
...
         def wrapper(self, *args, __remote_conn__=conn, __method_name__=method, **kw):
-            args = tuple(__remote_conn__.deliver(x) for x in args)
-            kw = {k: __remote_conn__.deliver(v) for k, v in kw.items()}
-            return getattr(self.__remote_end__, __method_name__)(*args, **kw)
+            args, kw = __remote_conn__.deliver(args, kw)
+            cache = object.__getattribute__(self, "__remote_methods__")
+            try:
+                remote = cache[__method_name__]
+            except KeyError:
+                cache[__method_name__] = remote = getattr(remote_cls, __method_name__)
```
Comment (anmyachev) [inline modin/experimental/cloud/rpyc_proxy.py:497]: You probably should do the same as in `__prepare__`

File: modin/experimental/cloud/rpyc_proxy.py [hunk 4]
```diff
@@ -12,34 +12,157 @@
...
+
+    def obtain_tuple(self, remote):
+        while True:
+            try:
+                remote = object.__getattribute__(remote, "__remote_end__")
+            except AttributeError:
+                break
+        return self._remote_tuplize(remote)
+
+    def sync_request(self, handler, *args):
```
Comment (anmyachev) [inline modin/experimental/cloud/rpyc_proxy.py:97]: please add docstring

File: modin/experimental/cloud/rpyc_proxy.py (hunk 1 above)
Comment (vnlitvinov) [reply modin/experimental/cloud/rpyc_proxy.py:51]: added a comment

File: modin/experimental/cloud/rpyc_proxy.py (hunk 2 above)
Comment (vnlitvinov) [reply modin/experimental/cloud/rpyc_proxy.py:53]: added a comment

File: modin/experimental/cloud/rpyc_proxy.py (hunk 3 above)
Comment (vnlitvinov) [reply modin/experimental/cloud/rpyc_proxy.py:497]: fixed, thx

File: modin/experimental/cloud/rpyc_proxy.py (hunk 4 above)
Comment (vnlitvinov) [reply modin/experimental/cloud/rpyc_proxy.py:97]: added

**Comment by anmyachev** [comment]
@vnlitvinov LGTM! Let's wait CI.

**Pull request merged**